### PR TITLE
MIGENG-70 RHVSavings rule and test

### PR DESCRIPTION
**Pull request by m-brophy**
https://jira.coreos.com/browse/MIGENG-70

### CODE REVIEW

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import org.jboss.xavier.analytics.pojo.output.RHVSavingsModel
+import org.jboss.xavier.analytics.pojo.output.RHVYearByYearCostsModel;
+
+dialect "java"
+agenda-group "RHVSavings"
+lock-on-active true
+
+rule "RHVSavingsRules"
+    when
+        report : InitialSavingsEstimationReportModel(rhvYearByYearCostsModel != null)
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
        report : InitialSavingsEstimationReportModel(
                rhvYearByYearCostsModel != null,
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        yearByYearCosts: RHVYearByYearCostsModel(
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            year1RhvBudgetFreedLowValue != null,
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
            rhvYearByYearCostsModel.year1RhvBudgetFreedLowValue != null,
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            year1RhvBudgetFreedLikelyValue != null,
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
            rhvYearByYearCostsModel.year1RhvBudgetFreedLikelyValue != null,
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            year1RhvBudgetFreedHighValue != null,
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
            rhvYearByYearCostsModel.year1RhvBudgetFreedHighValue != null,
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            year2RhvBudgetFreedLowValue != null,
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
            rhvYearByYearCostsModel.year2RhvBudgetFreedLowValue != null,
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            year2RhvBudgetFreedLikelyValue != null,
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
            rhvYearByYearCostsModel.year2RhvBudgetFreedLikelyValue != null,
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            year2RhvBudgetFreedHighValue != null,
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
            rhvYearByYearCostsModel.year2RhvBudgetFreedHighValue != null,
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            year3RhvBudgetFreedLowValue != null,
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
            rhvYearByYearCostsModel.year3RhvBudgetFreedLowValue != null,
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            year3RhvBudgetFreedLikelyValue != null,
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
            rhvYearByYearCostsModel.year3RhvBudgetFreedLikelyValue != null,
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+            year3RhvBudgetFreedHighValue != null
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
            rhvYearByYearCostsModel.year3RhvBudgetFreedHighValue != null
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        )
+    then
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl:29]:
```suggestion
    then
        RHVYearByYearCostsModel yearByYearCosts = report.getRhvYearByYearCostsModel();
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+        rhvSavingsModel.setRhvSaveLikelyValue(yearByYearCosts.getYear1RhvBudgetFreedLikelyValue() +
+                                              yearByYearCosts.getYear2RhvBudgetFreedLikelyValue() +
+                                              yearByYearCosts.getYear3RhvBudgetFreedLikelyValue());
+
+        rhvSavingsModel.setRhvSaveLowValue(yearByYearCosts.getYear1RhvBudgetFreedLowValue() +
+                                           yearByYearCosts.getYear2RhvBudgetFreedLowValue() +
+                                           yearByYearCosts.getYear3RhvBudgetFreedLowValue());
+
+        rhvSavingsModel.setRhvSaveFromValue(Math.floor(rhvSavingsModel.getRhvSaveLowValue()/100000)*100000);
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
```

File: src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+        rhvSavingsModel.setRhvSaveToValue(Math.floor(rhvSavingsModel.getRhvSaveHighValue()/100000)*100000);
```
Comment (mrizzi) [inline src/main/resources/org/jboss/xavier/analytics/rules/RHVSavings.drl]:
```suggestion
```

File: src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        rhvYearByYearCostsModel.setYear1RhvBudgetFreedLikelyValue(195951.0);
+        rhvYearByYearCostsModel.setYear1RhvBudgetFreedHighValue(370701.0);
+        rhvYearByYearCostsModel.setYear2RhvBudgetFreedLowValue(324262.0);
+        rhvYearByYearCostsModel.setYear2RhvBudgetFreedLikelyValue(499012.0);
+        rhvYearByYearCostsModel.setYear2RhvBudgetFreedHighValue(673762.0);
+        rhvYearByYearCostsModel.setYear3RhvBudgetFreedLowValue(254320.0);
+        rhvYearByYearCostsModel.setYear3RhvBudgetFreedLikelyValue(429070.0);
+        rhvYearByYearCostsModel.setYear3RhvBudgetFreedHighValue(603820.0);
+        // and add each object to the Map
+        facts.put("yearByYearCosts", rhvYearByYearCostsModel);
```
Comment (mrizzi) [inline src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java]:
```suggestion
```

File: src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+        rhvYearByYearCostsModel.setYear1RhvBudgetFreedLowValue(21201.0);
+        rhvYearByYearCostsModel.setYear1RhvBudgetFreedLikelyValue(195951.0);
+        rhvYearByYearCostsModel.setYear1RhvBudgetFreedHighValue(370701.0);
+        rhvYearByYearCostsModel.setYear2RhvBudgetFreedLowValue(324262.0);
+        rhvYearByYearCostsModel.setYear2RhvBudgetFreedLikelyValue(499012.0);
+        rhvYearByYearCostsModel.setYear2RhvBudgetFreedHighValue(673762.0);
+        rhvYearByYearCostsModel.setYear3RhvBudgetFreedLowValue(254320.0);
+        rhvYearByYearCostsModel.setYear3RhvBudgetFreedLikelyValue(429070.0);
+        rhvYearByYearCostsModel.setYear3RhvBudgetFreedHighValue(603820.0);
+        // and add each object to the Map
```
Comment (mrizzi) [inline src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java]:
```suggestion
```

File: src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+        facts.put("yearByYearCosts", rhvYearByYearCostsModel);
+
+        // another object needed as input for the test
+        EnvironmentModel environmentModel = new EnvironmentModel();
```
Comment (mrizzi) [inline src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java]:
```suggestion
```

File: src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+        InitialSavingsEstimationReportModel reportModel = new InitialSavingsEstimationReportModel();
+        reportModel.setRhvYearByYearCostsModel(rhvYearByYearCostsModel);
+        reportModel.setEnvironmentModel(environmentModel);
```
Comment (mrizzi) [inline src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java]:
```suggestion
```

File: src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+        // Check that the number of object is the right one (in this case, there must be just one report)
+        Assert.assertEquals(1, reports.size());
+
+        // Check that the object has exactly the fields that the rule tested should add/change
+        InitialSavingsEstimationReportModel report = reports.get(0);
+        RHVSavingsModel savingsModel = report.getRhvSavingsModel();
+        Assert.assertEquals(1648283, savingsModel.getRhvSaveHighValue(), 0);
+        Assert.assertEquals(1124033, savingsModel.getRhvSaveLikelyValue(), 0);
+        Assert.assertEquals(599783, savingsModel.getRhvSaveLowValue(), 0);
+        Assert.assertEquals(500000, savingsModel.getRhvSaveFromValue(), 0);
```
Comment (mrizzi) [inline src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java]:
```suggestion
```

File: src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+        Assert.assertEquals(1600000, savingsModel.getRhvSaveToValue(), 0);
```
Comment (mrizzi) [inline src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java]:
```suggestion
```

File: src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java [hunk 22]
```diff
@@ -0,0 +1,99 @@
+package org.jboss.xavier.analytics.test;
+
+import org.jboss.xavier.analytics.pojo.output.EnvironmentModel;
```
Comment (mrizzi) [inline src/test/java/org/jboss/xavier/analytics/test/RHVSavingsTest.java]:
```suggestion
```

**Review by mrizzi — CHANGES_REQUESTED** [review]
Hi @m-brophy,
i commented with my proposed changes to check that the inner `RHVYearByYearCostsModel` object is a good one.

**Comment by m-brophy** [comment]
retest

**Review by mrizzi — APPROVED** [review]
Thanks @m-brophy, it seems good to me now :+1:

**Pull request merged**